### DISCUSSION TITLE
Setting DNS servers?

### DISCUSSION THREAD

**Discussion by skulldogged** [pgj/freebsd-wifibox#112] [Q&A]
I'm using wifibox to handle wifi on freebsd and I want to change the DNS servers it uses (since the default one my router provides blocks some sites). I'm not exactly sure what the "correct" way of doing this is, I tried modifying udhcpd.conf but that doesn't seem to change anything. Is there something specific I need to do instead?

**Comment by pgj** [comment]
It should be managed in `udhcpd.conf`.  By the default, it has the following line:

```
opt     dns     8.8.8.8 8.8.4.4 %%DNS%%
```
where `8.8.8.8` and `8.8.4.4` are the default (Google) name servers that Wifibox offers and `%%DNS%%` is replaced with the ones that are received from the local network.

What do you see in `/etc/resolv.conf` on the FreeBSD host and what is expected?

**Reply by skulldogged** [reply]
My resolv.conf does look correct:

```
# Generated by resolvconf
nameserver 8.8.8.8
nameserver 8.8.4.4
nameserver 192.168.1.1
```

And this is my udhcpd.conf:

```
start           10.0.0.2
end             10.0.0.254
max_leases      64
interface       eth0
opt     subnet  255.255.255.0
opt     router  10.0.0.1
opt     dns     8.8.8.8
opt     dns     8.8.4.4
opt     dns     %%DNS%%
opt     lease   864000
```

But it seems that this isn't being respected - I know my router blocks autoconfig.kde.org and I still get the connection failure after following the steps in your other comment:
![image](https://github.com/user-attachments/assets/993fdc03-c876-4526-90eb-9df105e6bea7)

**Reply by pgj** [reply]
This is not necessarily caused by name resolution failure.  You can dig further to verify that, e.g.:

```console
$ host autoconfig.kde.org
autoconfig.kde.org is an alias for 1329036395.rsc.cdn77.org.
1329036395.rsc.cdn77.org has address 156.146.61.131
1329036395.rsc.cdn77.org has IPv6 address 2a02:6ea0:cb00::5
```

Or you can use `curl` to see how the connection builds up (or not):

```
curl -v https://autoconfig.kde.org/ocs/providers.xml
```

**Reply by skulldogged** [reply]
Looking with curl while using windows, autoconfig.kde.org fails with an SSL error with the default DNS and succeeds w/ both cloudflare and google DNS.

**Reply by pgj** [reply]
Interesting.  What is the SSL error?  Perhaps this reveals more of the actual cause.

**Reply by skulldogged** [reply]
Sorry for the late reply - the SSL error is definitely related to my wifi network, it's failing to resolve a mcafee URL:

```
curl -v https://autoconfig.kde.org/ocs/providers.xml
* Host autoconfig.kde.org:443 was resolved.
* IPv6: (none)
* IPv4: 109.61.91.230
*   Trying 109.61.91.230:443...
* Connected to autoconfig.kde.org (109.61.91.230) port 443
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256 / prime256v1 / rsaEncryption
* ALPN: server did not agree on a protocol. Uses default.
* Server certificate:
*  subject: C=US; ST=California; O=McAfee, LLC; CN=*.safezone.mcafee.com
*  start date: Jun 10 00:00:00 2024 GMT
*  expire date: Jun 10 23:59:59 2025 GMT
*  subjectAltName does not match autoconfig.kde.org
* SSL: no alternative certificate subject name matches target host name 'autoconfig.kde.org'
* Closing connection
* TLSv1.2 (OUT), TLS alert, close notify (256):
curl: (60) SSL: no alternative certificate subject name matches target host name 'autoconfig.kde.org'
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

**Reply by pgj** [reply]
For some reason, the original provider of `autconfig.kde.org`, CDN77 is blacklisted (?) by some McAfee antivirus product (which might be running on your router) and the request is getting relayed through a "safe zone" (`safezone.mcafee.com`).  But that is shipped with a different SSL certificate that will not match since that is for a site other than `autoconfig.kde.org`.

I would check the router of your local network to see if there is any configuration related to that.

**Reply by skulldogged** [reply]
I'm certain it is running on the router, as I get the same problem using the default DNS settings on any device. Unfortunately, I don't have access to the router's configuration, which is why I have been trying to instead change the DNS server wifibox uses.

**Comment by pgj** [comment]
One oddity I noticed about the format of `udhcpd.conf` is that it can take only 2 entries at maximum per line.  That is, the default setting above will not add anything that is going to be added in place of `%%DNS%%`.  Instead, the following format works much better:

```
opt dns 8.8.8.8
opt dns 8.8.4.4
opt dns %%DNS%%
```

Note that you would have to restart both the Wifibox guest and the `wifibox0` interface to have it propagated properly for the host's `resolv.conf`:

```console
# wifibox restart guest
# service netif restart wifibox0
```
